### PR TITLE
Issue 1570 - Fix issues with Reference search and add comprehensive tests

### PR DESCRIPTION
**Pull request by prb112**

### CODE REVIEW

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java [hunk 1]
```diff
@@ -529,6 +527,7 @@ private SqlQueryData processReferenceParm(Class<?> resourceType, QueryParameter
             // Handle query parm representing this name/value pair construct:
             // <code>{name}:{Resource Type} = {resource-id}</code>
             if (queryParm.getModifier() != null && queryParm.getModifier().equals(Modifier.TYPE)) {
+                System.out.println(queryParm.getType());
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java:530]: should it be a log statement?  or remove altogether?

Comment (prb112) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java:530]:
```suggestion
```

Comment (prb112) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/JDBCQueryBuilder.java]: fixed  - it was meant  for debugging, and then I wanted to remove it.  Thanks for the eagle eyes

File: fhir-search/src/main/java/com/ibm/fhir/search/reference/value/UrlHandlerImpl.java [hunk 2]
```diff
@@ -18,11 +18,11 @@
  * http://server/Patient/1
  * -> http://server/Patient/1
  * -> Patient/1
- * -> 1
+ * -> 1 (only when there is one target)
  */
 public class UrlHandlerImpl implements ParameterValueHandler {
 
-    private static final String REGEX = "/[A-z][a-z]{2,10}/[A-Za-z0-9\\-\\.]{1,64}$";
+    private static final String REGEX = "/([A-z][a-z]{2,36}/[A-Za-z0-9\\-\\.]{1,64})$";
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/reference/value/UrlHandlerImpl.java:25]: I assume the limit here is for performance reasons?
The longest resource type I found is 30 chars, but I still wonder if we should bump this to like 48 just to be catious.
Alternatively, we could check the actual list of resource names from the resourceTypes valueset and then use the one with max length, but thats probably overkill.

Comment (lmsurpre) [reply fhir-search/src/main/java/com/ibm/fhir/search/reference/value/UrlHandlerImpl.java:25]: we also have a related URL pattern in FHIRUtil.
```
sb.append(")\\/[A-Za-z0-9\\-\\.]{1,64}(\\/_history\\/[A-Za-z0-9\\-\\.]{1,64})?");
```

File: fhir-search/src/main/java/com/ibm/fhir/search/reference/value/TypeIdHandlerImpl.java [hunk 3]
```diff
@@ -6,18 +6,26 @@
...
 
 /**
  * The supported cases are:
  * Patient/1
  * -> http://server/Patient/1
+ * -> 1 (only when there is one target)
  */
 public class TypeIdHandlerImpl implements ParameterValueHandler {
+
+    private static final String REGEX = "^([A-z][a-z]{2,36}/[A-Za-z0-9\\-\\.]{1,64})$";
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/reference/value/TypeIdHandlerImpl.java:22]: same comment about erring on the side of caution on referencetype length
```suggestion
    private static final String REGEX = "^([A-z][a-z]{2,64}/[A-Za-z0-9\\-\\.]{1,64})$";
```

**Comment by lmsurpre** [comment]
Do the REGEX patterns in UrlHandlerImpl and TypeIdHandlerImpl fail when we get versioned references?
See https://www.hl7.org/fhir/search.html#versions for more info on these.
Not totally sure about our support for these to begin with, but probably want to at least consider them.

**Comment by prb112** [comment]
> Do the REGEX patterns in UrlHandlerImpl and TypeIdHandlerImpl fail when we get versioned references?
> See https://www.hl7.org/fhir/search.html#versions for more info on these.
> Not totally sure about our support for these to begin with, but probably want to at least consider them. I know at the least I opened #1273 recently and pointed Robin at it for consideration as part of his updated db schema.

There is CanonicalUrlHandlerImpl that handles that case, it splits it off. 
TypeId does not yet support version and the regex would fail. 

I did apply your review suggestions.

**Comment by lmsurpre** [comment]
> TypeId does not yet support version and the regex would fail.

Can we add this to the end of the REGEX for that one: 
`(\\/_history\\/[A-Za-z0-9\\-\\.]{1,64})?`

And then maybe a single test which exercises it?

**Comment by prb112** [comment]
> > TypeId does not yet support version and the regex would fail.
> 
> Can we add this to the end of the REGEX (but before the `$`) for both that one and for UrlHandlerImpl:
> `(\\/_history\\/[A-Za-z0-9\\-\\.]{1,64})?`
> 
> And then maybe a single test which exercises it for each?

Sure - that's no problemo

**Comment by prb112** [comment]
> > TypeId does not yet support version and the regex would fail.
> 
> Can we add this to the end of the REGEX (but before the `$`) for both that one and for UrlHandlerImpl:
> `(\\/_history\\/[A-Za-z0-9\\-\\.]{1,64})?`
> 
> And then maybe a single test which exercises it for each?

Updated

**Review by lmsurpre — APPROVED** [review]
LGTM

**Pull request merged**